src/main/ngapp/accio/src/app/editor/magic-wand.service.ts

```diff
@@ -0,0 +1,159 @@
+                         pixelCoord: Array<number>, tolerance: number,
+                         visited: Set<number>): boolean {
+    let isStillMask: boolean = true;
+
+    // Preface; check if pixel is valid (indexing errs and repeat values)
+    let isValid: boolean = this.getIsValid(imgData, pixelCoord, visited);
+    if (!isValid) {
+      // Automatically not in mask b/c failed vailidity test
+      return !isStillMask;
+    }
+
+    // Get array of attributes of current pixel
+    let curX: number = pixelCoord[0];
+    let curY: number = pixelCoord[1];
+    let curPixel: Array<number> = this.dataArrayToRGBA(imgData, curX, curY);
+
+    // All attributes of the pixel (R,G,B, and A) must be within tolerance level
+    for (let i = 0; i < 4; i++) {
+      let upperTolerance: boolean = curPixel[i] > originalPixel[i] + tolerance;
+      let lowerTolerance: boolean = curPixel[i] < originalPixel[i] - tolerance;
+      let failsTolerance: boolean = upperTolerance || lowerTolerance;
+      if (failsTolerance) {
+        return !isStillMask;
+      }
+    }
+
+    return isStillMask;
+  }
+
+  // Check if @pixelCoord is in bounds and makes sure it's not a repeat coord
+  getIsValid(imgData: ImageData, pixelCoord: Array<number>, visited: Set<number>): boolean {
+    let curX: number = pixelCoord[0];
+    let curY: number = pixelCoord[1];
+
+    // Unpack imgData for readability
+    let data: Uint8ClampedArray = imgData.data;
+    let imgWidth: number = imgData.width;
+    let imgHeight: number = imgData.height;
+
+    let isValid: boolean = true;
+    // Check bounds of indexing
+    let yOutOfBounds: boolean = curY < 0 || curY > imgHeight - 1;
+    let xOutOfBounds: boolean = curX < 0 || curX > imgWidth - 1;
+    if (yOutOfBounds || xOutOfBounds) {
+      return !isValid;
+    }
+    // Do not push repeat coords to heap
+    let indexAsDataArray: number = this.coordToDataArrayIndicies(curX, curY, imgWidth)[0];
+    if (visited.has(indexAsDataArray)) {
+      return !isValid;
+    }
+    visited.add(indexAsDataArray);
+
+    return isValid;
+  }
+
+  // Return pixel attributes of @imgData at [@xCoord, @yCoord] as Array<number>
+  dataArrayToRGBA(imgData: ImageData, xCoord: number, yCoord: number): Array<number> {
+    // Unpack imgData for readability
+    let data: Uint8ClampedArray = imgData.data;
+    let imgWidth: number = imgData.width;
+    let imgHeight: number = imgData.height;
+
+    // Pixel attributes in imgData are organized adjacently in a 1-D array
+    let pixelIndex: number = (xCoord + (imgWidth * yCoord)) * 4;
+    let red: number = data[pixelIndex];
+    let green: number = data[pixelIndex + 1];
+    let blue: number = data[pixelIndex + 2];
+    let alpha: number = data[pixelIndex + 3];
+    // Store original pixel's attributes
+    let pixel: Array<number> = [red, green, blue, alpha];
+    return pixel;
+  }
+
+  // Convert coord [@x, @y] to indexing style of imageData.data
```

Comment:
might add in the comments that this assumes x & y are coordinates in the imageData

Reply:
I don't think that necessarily applies here. This is only a general helper function to convert from a 2-D array to a 1-D array indexing style, and so it's purpose is not explicitly to interact with an ImageData object on it's own.

Reply:
Instead, I will change comment to not suggest it is connected to ImageData but instead  DataArray